framework/include/utils/MooseUtils.h

```diff
@@ -57,15 +58,13 @@ class MultiMooseEnum;
 namespace MooseUtils
 {
 
-std::string pathjoin(const std::string & s);
+std::filesystem::path pathjoin(const std::filesystem::path & p);
 
 template <typename... Args>
-std::string
-pathjoin(const std::string & s, Args... args)
+std::filesystem::path
+pathjoin(const std::filesystem::path & p, Args... args)
```

Comment:
I can't believe I just got schooled on C++ here.  I know you can't implicitly convert a std::string` (e.g. from `getExecutablePath()` below) to a `std::filesystem::path &`, so I couldn't see how this compiled, but your argument is a *const* `path &`, and either I'd forgotten or I'd never learned that you could convert a temporary to get a temporary-compatible reference.

Reply:
Yeah, that's right. Where this blows up are our derived string types. So I may need to explicitly create a std::string temporary as an intermediate, like in the last commit.

Reply:
Oh, I get it - C++ will automatically chain together a user-defined conversion with a couple basic conversions before-and-after, but it won't implicitly chain together multiple user-defined conversions at once.

If this were in libMesh I'd have just kept the `pathjoin(string, args)` overload, as the only guaranteed backward compatibility, but marked it deprecated for a while before removing it.  But MOOSE has enough CI coverage of downstream apps that I don't see any problem with just removing it right away so long as the apps tests all pass.  This is entirely a compile-time rather than run-time issue.

Reply:
Yeah, there's a problem though with that overload. I believe if you keep it you end up with an ambiguous call, because with a `std::string` argument you can call both the path and the string version. I can double check what exactly is going on there, but I went through a couple of iterations here and some didn't compile due to the ambiguity. Hence I went with the templates solution. Now what I really want are C++ _concepts_, which would let me define a set of acceptable types that can represent a path (`std::string`, `FileName`, `std::filesystem::path`).